File: test/language/templates/cpp/InstantiateTypeAsParameterTest.cpp
#include "gtest/gtest.h"

#include "templates/instantiate_type_as_parameter/InstantiateTypeAsParameter.h"

#include "zserio/RebindAlloc.h"

namespace templates
{
namespace instantiate_type_as_parameter
{

using allocator_type = InstantiateTypeAsParameter::allocator_type;
template <typename T>
using vector_type = std::vector<T, zserio::RebindAlloc<allocator_type, T>>;

TEST(InstantiateTypeAsParameterTest, readWrite)
{
    InstantiateTypeAsParameter instantiateTypeAsParameter{
            P32{2}, Parameterized_P32{vector_type<uint32_t>{13, 42}}
    };

    zserio::BitBuffer bitBuffer = zserio::BitBuffer(1024 * 8);
    zserio::BitStreamWriter writer(bitBuffer);
    instantiateTypeAsParameter.write(writer);

    zserio::BitStreamReader reader(writer.getWriteBuffer(), writer.getBitPosition(), zserio::BitsTag());
    InstantiateTypeAsParameter readInstantiateTypeAsParameter(reader);

    ASSERT_TRUE(instantiateTypeAsParameter == readInstantiateTypeAsParameter);
}

} // namespace instantiate_type_as_parameter
} // namespace templates

File: test/language/templates/python/InstantiateTypeAsParameterTest.py
import unittest
import zserio

from testutils import getZserioApi

class InstantiateTypeAsParameterTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "templates.zs").instantiate_type_as_parameter

    def testReadWrite(self):
        param = self.api.P32(2)
        instantiateTypeAsParameter = self.api.InstantiateTypeAsParameter(
            param, self.api.Parameterized_P32(param, [13, 42])
        )

        writer = zserio.BitStreamWriter()
        instantiateTypeAsParameter.write(writer)
        reader = zserio.BitStreamReader(writer.getByteArray(), writer.getBitPosition())
        readInstantiateTypeAsParameter = self.api.InstantiateTypeAsParameter()
        readInstantiateTypeAsParameter.read(reader)
        self.assertEqual(instantiateTypeAsParameter, readInstantiateTypeAsParameter)
